File: frameworks/shared/php/src/HealthCheckManager.php
<?php

namespace VelocityBench\HealthCheck;

/**
 * Health check manager for PHP frameworks.
 */
class HealthCheckManager
{
    private float $startTime;
    private ?\PDO $database = null;
    private array $cache = [];
    private int $cacheTtlMs = 5000; // 5 seconds

    public function __construct(
        private HealthCheckConfig $config
    ) {
        $this->startTime = microtime(true);
    }

    /**
     * Set database connection for health checks.
     */
    public function withDatabase(\PDO $pdo): self
    {
        $this->database = $pdo;
        return $this;
    }

    /**
     * Execute a health check probe.
     */
    public function probe(string $probeType): HealthCheckResponse
    {
        $probe = ProbeType::fromString($probeType);

        // Check cache
        $cached = $this->getCachedResult($probe);
        if ($cached !== null) {
            return $cached;
        }

        // Execute probe
        $result = match ($probe) {
            ProbeType::LIVENESS => $this->livenessProbe(),
            ProbeType::READINESS => $this->readinessProbe(),
            ProbeType::STARTUP => $this->startupProbe(),
        };

        // Cache result
        $this->cacheResult($probe, $result);

        return $result;
    }

    /**
     * Get cached result if still valid.
     */
    private function getCachedResult(ProbeType $probeType): ?HealthCheckResponse
    {
        $key = $probeType->value;

        if (!isset($this->cache[$key])) {
            return null;
        }

        $cached = $this->cache[$key];
        $age = (microtime(true) * 1000) - $cached['timestamp'];

        if ($age < $this->cacheTtlMs) {
            return $cached['result'];
        }

        return null;
    }

    /**
     * Cache a health check result.
     */
    private function cacheResult(ProbeType $probeType, HealthCheckResponse $result): void
    {
        $this->cache[$probeType->value] = [
            'result' => $result,
            'timestamp' => microtime(true) * 1000,
        ];
    }

    /**
     * Liveness probe: Is the process alive?
     */
    private function livenessProbe(): HealthCheckResponse
    {
        $checks = [
            'memory' => $this->checkMemory(),
        ];

        $overallStatus = $this->computeOverallStatus($checks);

        return new HealthCheckResponse(
            status: $overallStatus,
            timestamp: $this->getTimestamp(),
            uptimeMs: $this->getUptimeMs(),
            version: $this->config->version,
            service: $this->config->serviceName,
            environment: $this->config->environment,
            probeType: ProbeType::LIVENESS,
            checks: $checks
        );
    }

    /**
     * Readiness probe: Can the service handle traffic?
     */
    private function readinessProbe(): HealthCheckResponse
    {
        $checks = [];

        // Database check
        if ($this->database !== null) {
            $checks['database'] = $this->checkDatabase();
        }

        // Memory check
        $checks['memory'] = $this->checkMemory();

        $overallStatus = $this->computeOverallStatus($checks);

        return new HealthCheckResponse(
            status: $overallStatus,
            timestamp: $this->getTimestamp(),
            uptimeMs: $this->getUptimeMs(),
            version: $this->config->version,
            service: $this->config->serviceName,
            environment: $this->config->environment,
            probeType: ProbeType::READINESS,
            checks: $checks
        );
    }

    /**
     * Startup probe: Has initialization completed?
     */
    private function startupProbe(): HealthCheckResponse
    {
        $checks = [];

        // Database check
        if ($this->database !== null) {
            $checks['database'] = $this->checkDatabase();
        }

        // Warmup check
        $checks['warmup'] = $this->checkWarmup();

        // Memory check
        $checks['memory'] = $this->checkMemory();

        $overallStatus = $this->computeOverallStatus($checks);

        return new HealthCheckResponse(
            status: $overallStatus,
            timestamp: $this->getTimestamp(),
            uptimeMs: $this->getUptimeMs(),
            version: $this->config->version,
            service: $this->config->serviceName,
            environment: $this->config->environment,
            probeType: ProbeType::STARTUP,
            checks: $checks
        );
    }

    /**
     * Check database connectivity.
     */
    private function checkDatabase(): HealthCheck
    {
        $start = microtime(true);

        try {
            // Execute simple query with timeout
            $this->database->setAttribute(\PDO::ATTR_TIMEOUT, 3);
            $stmt = $this->database->query('SELECT 1');
            $stmt->fetch();

            $responseTime = (microtime(true) - $start) * 1000;

            return (new HealthCheck(HealthStatus::UP))
                ->withResponseTime($responseTime)
                ->withData('pool_size', 0)
                ->withData('pool_available', 0);

        } catch (\PDOException $e) {
            return (new HealthCheck(HealthStatus::DOWN))
                ->withError('Database connection error: ' . $e->getMessage());
        }
    }

    /**
     * Check memory usage.
     */
    private function checkMemory(): HealthCheck
    {
        $usedMemory = memory_get_usage(true);
        $peakMemory = memory_get_peak_usage(true);
        $memoryLimit = $this->getMemoryLimit();

        $usedMb = $usedMemory / (1024 * 1024);
        $totalMb = $memoryLimit / (1024 * 1024);
        $utilization = ($usedMemory / $memoryLimit) * 100;

        $check = (new HealthCheck(HealthStatus::UP))
            ->withData('used_mb', round($usedMb, 2))
            ->withData('total_mb', round($totalMb, 2))
            ->withData('utilization_percent', round($utilization, 2));

        if ($utilization > 90.0) {
            $check->status = HealthStatus::DEGRADED;
            $check->withWarning(sprintf('Critical memory usage (%.1f%%)', $utilization));
        } elseif ($utilization > 80.0) {
            $check->withWarning(sprintf('High memory usage (%.1f%%)', $utilization));
        }

        return $check;
    }

    /**
     * Check if warmup period has completed.
     */
    private function checkWarmup(): HealthCheck
    {
        $uptimeMs = $this->getUptimeMs();

        if ($uptimeMs < $this->config->startupDurationMs) {
            $progress = ($uptimeMs / $this->config->startupDurationMs) * 100;
            return (new HealthCheck(HealthStatus::IN_PROGRESS))
                ->withInfo(sprintf('Warming up (%.0f%% complete)', $progress))
                ->withData('progress_percent', round($progress, 2))
                ->withData('uptime_ms', $uptimeMs)
                ->withData('target_ms', $this->config->startupDurationMs);
        }

        return (new HealthCheck(HealthStatus::UP))
            ->withInfo('Warmup complete');
    }

    /**
     * Compute overall health status from individual checks.
     */
    private function computeOverallStatus(array $checks): HealthStatus
    {
        $hasDown = false;
        $hasInProgress = false;
        $hasDegraded = false;

        foreach ($checks as $check) {
            match ($check->status) {
                HealthStatus::DOWN => $hasDown = true,
                HealthStatus::IN_PROGRESS => $hasInProgress = true,
                HealthStatus::DEGRADED => $hasDegraded = true,
                default => null,
            };
        }

        if ($hasDown) {
            return HealthStatus::DOWN;
        } elseif ($hasInProgress) {
            return HealthStatus::IN_PROGRESS;
        } elseif ($hasDegraded) {
            return HealthStatus::DEGRADED;
        }

        return HealthStatus::UP;
    }

    /**
     * Get service uptime in milliseconds.
     */
    private function getUptimeMs(): int
    {
        return (int) ((microtime(true) - $this->startTime) * 1000);
    }

    /**
     * Get current timestamp in ISO 8601 format.
     */
    private function getTimestamp(): string
    {
        return (new \DateTime('now', new \DateTimeZone('UTC')))
            ->format('Y-m-d\TH:i:s.v\Z');
    }

    /**
     * Get PHP memory limit in bytes.
     */
    private function getMemoryLimit(): int
    {
        $memoryLimit = ini_get('memory_limit');

        if ($memoryLimit === '-1') {
            // Unlimited memory, return a large value
            return PHP_INT_MAX;
        }

        // Parse memory limit (e.g., "128M", "1G")
        $value = (int) $memoryLimit;
        $unit = strtoupper(substr($memoryLimit, -1));

        return match ($unit) {
            'G' => $value * 1024 * 1024 * 1024,
            'M' => $value * 1024 * 1024,
            'K' => $value * 1024,
            default => $value,
        };
    }
}

File: frameworks/shared/ruby/lib/velocitybench/healthcheck/health_check_manager.rb
# frozen_string_literal: true

require 'time'
require 'json'

module VelocityBench
  module HealthCheck
    # Health check manager for Ruby frameworks.
    class HealthCheckManager
      CACHE_TTL_MS = 5000 # 5 seconds

      def initialize(config)
        @config = config
        @start_time = Time.now
        @database = nil
        @cache = {}
      end

      # Set database connection for health checks.
      def with_database(connection)
        @database = connection
        self
      end

      # Execute a health check probe.
      def probe(probe_type)
        probe = ProbeType.from_string(probe_type)

        # Check cache
        cached = get_cached_result(probe)
        return cached if cached

        # Execute probe
        result = case probe
                 when ProbeType::LIVENESS
                   liveness_probe
                 when ProbeType::READINESS
                   readiness_probe
                 when ProbeType::STARTUP
                   startup_probe
                 end

        # Cache result
        cache_result(probe, result)

        result
      end

      private

      # Get cached result if still valid.
      def get_cached_result(probe_type)
        return nil unless @cache[probe_type]

        cached = @cache[probe_type]
        age_ms = (Time.now.to_f - cached[:timestamp]) * 1000

        return cached[:result] if age_ms < CACHE_TTL_MS

        nil
      end

      # Cache a health check result.
      def cache_result(probe_type, result)
        @cache[probe_type] = {
          result: result,
          timestamp: Time.now.to_f
        }
      end

      # Liveness probe: Is the process alive?
      def liveness_probe
        checks = {
          'memory' => check_memory
        }

        overall_status = compute_overall_status(checks)

        HealthCheckResponse.new(
          status: overall_status,
          timestamp: get_timestamp,
          uptime_ms: get_uptime_ms,
          version: @config.version,
          service: @config.service_name,
          environment: @config.environment,
          probe_type: ProbeType::LIVENESS,
          checks: checks
        )
      end

      # Readiness probe: Can the service handle traffic?
      def readiness_probe
        checks = {}

        # Database check
        checks['database'] = check_database if @database

        # Memory check
        checks['memory'] = check_memory

        overall_status = compute_overall_status(checks)

        HealthCheckResponse.new(
          status: overall_status,
          timestamp: get_timestamp,
          uptime_ms: get_uptime_ms,
          version: @config.version,
          service: @config.service_name,
          environment: @config.environment,
          probe_type: ProbeType::READINESS,
          checks: checks
        )
      end

      # Startup probe: Has initialization completed?
      def startup_probe
        checks = {}

        # Database check
        checks['database'] = check_database if @database

        # Warmup check
        checks['warmup'] = check_warmup

        # Memory check
        checks['memory'] = check_memory

        overall_status = compute_overall_status(checks)

        HealthCheckResponse.new(
          status: overall_status,
          timestamp: get_timestamp,
          uptime_ms: get_uptime_ms,
          version: @config.version,
          service: @config.service_name,
          environment: @config.environment,
          probe_type: ProbeType::STARTUP,
          checks: checks
        )
      end

      # Check database connectivity.
      def check_database
        start_time = Time.now

        begin
          # Execute simple query
          @database.exec('SELECT 1')

          response_time = (Time.now - start_time) * 1000

          HealthCheck.new(HealthStatus::UP)
            .with_response_time(response_time.round(2))
            .with_data('pool_size', 0)
            .with_data('pool_available', 0)

        rescue StandardError => e
          HealthCheck.new(HealthStatus::DOWN)
            .with_error("Database connection error: #{e.message}")
        end
      end

      # Check memory usage.
      def check_memory
        # Get process memory usage (RSS)
        pid = Process.pid
        memory_kb = `ps -o rss= -p #{pid}`.strip.to_i
        used_mb = memory_kb / 1024.0

        # Estimate total available memory (simplified)
        total_mb = 2048.0 # Default assumption
        utilization = (used_mb / total_mb) * 100

        check = HealthCheck.new(HealthStatus::UP)
          .with_data('used_mb', used_mb.round(2))
          .with_data('total_mb', total_mb.round(2))
          .with_data('utilization_percent', utilization.round(2))

        if utilization > 90.0
          check.status = HealthStatus::DEGRADED
          check.with_warning(format('Critical memory usage (%.1f%%)', utilization))
        elsif utilization > 80.0
          check.with_warning(format('High memory usage (%.1f%%)', utilization))
        end

        check
      end

      # Check if warmup period has completed.
      def check_warmup
        uptime_ms = get_uptime_ms

        if uptime_ms < @config.startup_duration_ms
          progress = (uptime_ms.to_f / @config.startup_duration_ms) * 100
          HealthCheck.new(HealthStatus::IN_PROGRESS)
            .with_info(format('Warming up (%.0f%% complete)', progress))
            .with_data('progress_percent', progress.round(2))
            .with_data('uptime_ms', uptime_ms)
            .with_data('target_ms', @config.startup_duration_ms)
        else
          HealthCheck.new(HealthStatus::UP)
            .with_info('Warmup complete')
        end
      end

      # Compute overall health status from individual checks.
      def compute_overall_status(checks)
        has_down = false
        has_in_progress = false
        has_degraded = false

        checks.each_value do |check|
          case check.status
          when HealthStatus::DOWN
            has_down = true
          when HealthStatus::IN_PROGRESS
            has_in_progress = true
          when HealthStatus::DEGRADED
            has_degraded = true
          end
        end

        return HealthStatus::DOWN if has_down
        return HealthStatus::IN_PROGRESS if has_in_progress
        return HealthStatus::DEGRADED if has_degraded

        HealthStatus::UP
      end

      # Get service uptime in milliseconds.
      def get_uptime_ms
        ((Time.now - @start_time) * 1000).to_i
      end

      # Get current timestamp in ISO 8601 format.
      def get_timestamp
        Time.now.utc.iso8601(3)
      end
    end
  end
end
